File: MiniDB_Projects/Team_Alpha/src/minidb/sql/Parser.java
package minidb.sql;

import minidb.common.Types.*;
import java.util.*;

/**
 * A hand-written recursive-descent parser for MiniDB's SQL subset.
 *
 * Supported grammar (case-insensitive keywords):
 *   CREATE TABLE t (col TYPE [PRIMARY KEY], ...)
 *   CREATE INDEX ON t (col)
 *   INSERT INTO t VALUES (v1, v2, ...)
 *   SELECT cols FROM t1 [, t2] [WHERE pred [AND pred ...]]
 *   DELETE FROM t [WHERE pred]
 *   BEGIN | COMMIT | ABORT
 *
 * A predicate is: <col> <op> <literal|col>, where col may be qualified t.col.
 *
 * The parser turns text into an Ast.Stmt; it does no semantic checking — that
 * happens in the executor against the catalog.
 */
public final class Parser {
    private final List<String> toks;
    private int pos = 0;

    public Parser(String sql) { this.toks = tokenize(sql); }

    public Ast.Stmt parse() {
        String kw = peek().toUpperCase();
        switch (kw) {
            case "CREATE": return parseCreate();
            case "INSERT": return parseInsert();
            case "SELECT": return parseSelect();
            case "DELETE": return parseDelete();
            case "BEGIN":  next(); return new Ast.Begin();
            case "COMMIT": next(); return new Ast.Commit();
            case "ABORT": case "ROLLBACK": next(); return new Ast.Abort();
            default: throw err("Unknown statement: " + kw);
        }
    }

    // ---- CREATE TABLE / CREATE INDEX ----
    private Ast.Stmt parseCreate() {
        expect("CREATE");
        if (peek().equalsIgnoreCase("INDEX")) {
            next(); expect("ON");
            Ast.CreateIndex ci = new Ast.CreateIndex();
            ci.table = next();
            expect("(");
            ci.column = next();
            expect(")");
            return ci;
        }
        expect("TABLE");
        Ast.CreateTable ct = new Ast.CreateTable();
        ct.table = next();
        expect("(");
        while (true) {
            String col = next();
            String type = next().toUpperCase();
            ColType ct2 = type.startsWith("INT") ? ColType.INT : ColType.STRING;
            ct.columns.add(new Column(col, ct2));
            if (peek().equalsIgnoreCase("PRIMARY")) {
                next(); expect("KEY");
                ct.primaryKey = col;
            }
            if (peek().equals(",")) { next(); continue; }
            break;
        }
        expect(")");
        return ct;
    }

    // ---- INSERT ----
    private Ast.Stmt parseInsert() {
        expect("INSERT"); expect("INTO");
        Ast.Insert ins = new Ast.Insert();
        ins.table = next();
        expect("VALUES"); expect("(");
        while (true) {
            ins.values.add(literal(next()));
            if (peek().equals(",")) { next(); continue; }
            break;
        }
        expect(")");
        return ins;
    }

    // ---- DELETE ----
    private Ast.Stmt parseDelete() {
        expect("DELETE"); expect("FROM");
        Ast.Delete d = new Ast.Delete();
        d.table = next();
        if (!atEnd() && peek().equalsIgnoreCase("WHERE")) {
            next();
            d.where = parsePredicate();
        }
        return d;
    }

    // ---- SELECT ----
    private Ast.Stmt parseSelect() {
        expect("SELECT");
        Ast.Select s = new Ast.Select();
        if (peek().equals("*")) { next(); s.projections.add("*"); }
        else {
            while (true) {
                s.projections.add(next());
                if (peek().equals(",")) { next(); continue; }
                break;
            }
        }
        expect("FROM");
        while (true) {
            s.fromTables.add(next());
            if (peek().equals(",")) { next(); continue; }
            break;
        }
        if (!atEnd() && peek().equalsIgnoreCase("WHERE")) {
            next();
            while (true) {
                s.wheres.add(parsePredicate());
                if (!atEnd() && peek().equalsIgnoreCase("AND")) { next(); continue; }
                break;
            }
        }
        return s;
    }

    private Ast.Predicate parsePredicate() {
        Ast.Predicate p = new Ast.Predicate();
        String left = next();
        String[] lq = qualify(left);
        p.leftTable = lq[0]; p.leftCol = lq[1];
        p.op = next();
        String right = next();
        if (isLiteralToken(right)) {
            p.rhsIsColumn = false;
            p.rhsValue = literal(right);
        } else {
            p.rhsIsColumn = true;
            String[] rq = qualify(right);
            p.rightTable = rq[0]; p.rightCol = rq[1];
        }
        return p;
    }

    private String[] qualify(String tok) {
        int dot = tok.indexOf('.');
        if (dot >= 0) return new String[]{ tok.substring(0, dot), tok.substring(dot + 1) };
        return new String[]{ null, tok };
    }

    private boolean isLiteralToken(String t) {
        if (t.startsWith("'")) return true;
        if (t.isEmpty()) return false;
        char c = t.charAt(0);
        return Character.isDigit(c) || c == '-';
    }

    private Object literal(String t) {
        if (t.startsWith("'")) return t.substring(1, t.length() - 1); // strip quotes
        try { return Integer.parseInt(t); } catch (NumberFormatException e) { return t; }
    }

    // ---- tokenizer ----
    private static List<String> tokenize(String sql) {
        List<String> out = new ArrayList<>();
        int i = 0, n = sql.length();
        while (i < n) {
            char c = sql.charAt(i);
            if (Character.isWhitespace(c)) { i++; continue; }
            if (c == ';') { i++; continue; }
            if (c == '\'') { // string literal
                int j = i + 1;
                StringBuilder sb = new StringBuilder("'");
                while (j < n && sql.charAt(j) != '\'') sb.append(sql.charAt(j++));
                sb.append('\'');
                out.add(sb.toString());
                i = j + 1;
                continue;
            }
            if ("(),*".indexOf(c) >= 0) { out.add(String.valueOf(c)); i++; continue; }
            if ("<>=!".indexOf(c) >= 0) { // operators, possibly two-char
                int j = i + 1;
                if (j < n && "<>=!".indexOf(sql.charAt(j)) >= 0) { out.add(sql.substring(i, j + 1)); i = j + 1; }
                else { out.add(String.valueOf(c)); i++; }
                continue;
            }
            int j = i; // identifier / number / keyword
            while (j < n && !Character.isWhitespace(sql.charAt(j))
                    && "(),*<>=!;'".indexOf(sql.charAt(j)) < 0) j++;
            out.add(sql.substring(i, j));
            i = j;
        }
        return out;
    }

    // ---- token cursor helpers ----
    private String peek() { return atEnd() ? "" : toks.get(pos); }
    private String next() { return toks.get(pos++); }
    private boolean atEnd() { return pos >= toks.size(); }
    private void expect(String kw) {
        String t = next();
        if (!t.equalsIgnoreCase(kw)) throw err("Expected '" + kw + "' but got '" + t + "'");
    }
    private RuntimeException err(String m) { return new RuntimeException("Parse error: " + m); }
}

File: MiniDB_Projects/Team_IndexIQ/src/parser/parser.cpp
#include "parser/parser.h"
#include <stdexcept>

Parser::Parser(std::vector<Token> tokens) : tokens_(std::move(tokens)) {}

Token Parser::consume(TokenType expected) {
    if (cur().type != expected)
        throw std::runtime_error("Parse error: unexpected token '" + cur().text + "'");
    return tokens_[pos_++];
}

Statement Parser::parse() {
    if (match(TokenType::EXPLAIN)) return parse_select(true);
    if (check(TokenType::SELECT))  { pos_++; return parse_select(false); }
    if (check(TokenType::INSERT))  { pos_++; return parse_insert(); }
    if (check(TokenType::DELETE))  { pos_++; return parse_delete(); }
    if (check(TokenType::CREATE))  { pos_++; return parse_create(); }
    if (match(TokenType::BEGIN))   return BeginStmt{};
    if (match(TokenType::COMMIT))  return CommitStmt{};
    if (match(TokenType::ROLLBACK))return RollbackStmt{};
    throw std::runtime_error("Unknown statement: " + cur().text);
}

Statement Parser::parse_select(bool explain) {
    if (explain) consume(TokenType::SELECT);

    SelectStmt stmt;
    stmt.explain = explain;

    if (match(TokenType::STAR)) {
        stmt.cols = {"*"};
    } else {
        stmt.cols.push_back(consume(TokenType::IDENTIFIER).text);
        while (match(TokenType::COMMA))
            stmt.cols.push_back(consume(TokenType::IDENTIFIER).text);
    }

    consume(TokenType::FROM);
    stmt.table = consume(TokenType::IDENTIFIER).text;

    if (match(TokenType::JOIN)) {
        stmt.join_table = consume(TokenType::IDENTIFIER).text;
        consume(TokenType::ON);
        std::string lt = consume(TokenType::IDENTIFIER).text;
        consume(TokenType::DOT);
        std::string lc = consume(TokenType::IDENTIFIER).text;
        consume(TokenType::EQ);
        std::string rt = consume(TokenType::IDENTIFIER).text;
        consume(TokenType::DOT);
        std::string rc = consume(TokenType::IDENTIFIER).text;
        stmt.join_left_col  = lt + "." + lc;
        stmt.join_right_col = rt + "." + rc;
    }

    if (match(TokenType::WHERE))
        stmt.where = parse_expr();

    return stmt;
}

Statement Parser::parse_insert() {
    consume(TokenType::INTO);
    InsertStmt stmt;
    stmt.table = consume(TokenType::IDENTIFIER).text;
    consume(TokenType::VALUES);
    consume(TokenType::LPAREN);
    auto read_val = [&]() -> std::string {
        if (check(TokenType::NUMBER))     return consume(TokenType::NUMBER).text;
        if (check(TokenType::STRING))     return consume(TokenType::STRING).text;
        if (check(TokenType::IDENTIFIER)) return consume(TokenType::IDENTIFIER).text;
        throw std::runtime_error("Expected value in INSERT");
    };
    stmt.values.push_back(read_val());
    while (match(TokenType::COMMA)) stmt.values.push_back(read_val());
    consume(TokenType::RPAREN);
    return stmt;
}

Statement Parser::parse_delete() {
    consume(TokenType::FROM);
    DeleteStmt stmt;
    stmt.table = consume(TokenType::IDENTIFIER).text;
    if (match(TokenType::WHERE)) stmt.where = parse_expr();
    return stmt;
}

Statement Parser::parse_create() {
    consume(TokenType::TABLE);
    CreateStmt stmt;
    stmt.table = consume(TokenType::IDENTIFIER).text;
    consume(TokenType::LPAREN);
    auto read_col = [&]() {
        ColDef c;
        c.name = consume(TokenType::IDENTIFIER).text;
        c.type = consume(TokenType::IDENTIFIER).text;
        for (char& ch : c.type) ch = std::toupper(ch);
        stmt.columns.push_back(c);
    };
    read_col();
    while (match(TokenType::COMMA)) read_col();
    consume(TokenType::RPAREN);
    return stmt;
}

Expression* Parser::parse_expr() {
    auto* left = parse_and();
    while (match(TokenType::OR)) {
        auto* right = parse_and();
        left = new BinaryExpr("OR", left, right);
    }
    return left;
}

Expression* Parser::parse_and() {
    auto* left = parse_comparison();
    while (match(TokenType::AND)) {
        auto* right = parse_comparison();
        left = new BinaryExpr("AND", left, right);
    }
    return left;
}

Expression* Parser::parse_comparison() {
    auto* left = parse_primary();
    std::string op;
    if      (match(TokenType::EQ))  op = "=";
    else if (match(TokenType::NEQ)) op = "!=";
    else if (match(TokenType::GT))  op = ">";
    else if (match(TokenType::LT))  op = "<";
    else if (match(TokenType::GTE)) op = ">=";
    else if (match(TokenType::LTE)) op = "<=";
    else return left;
    auto* right = parse_primary();
    return new BinaryExpr(op, left, right);
}

Expression* Parser::parse_primary() {
    if (check(TokenType::NUMBER))
        return new Literal(consume(TokenType::NUMBER).text);
    if (check(TokenType::STRING))
        return new Literal(consume(TokenType::STRING).text);
    if (check(TokenType::IDENTIFIER)) {
        std::string name = consume(TokenType::IDENTIFIER).text;
        if (match(TokenType::DOT)) {
            std::string col = consume(TokenType::IDENTIFIER).text;
            return new ColumnRef(name, col);
        }
        return new ColumnRef(name);
    }
    if (match(TokenType::LPAREN)) {
        auto* e = parse_expr();
        consume(TokenType::RPAREN);
        return e;
    }
    throw std::runtime_error("Expected expression, got: " + cur().text);
}
